pkg/table/tables/tables.go

```diff
@@ -581,18 +595,18 @@ func (t *TableCommon) UpdateRecord(ctx context.Context, sctx table.MutateContext
 			checksumData = t.appendPublicColForChecksum(sctx, h, checksumData, col.ToInfo(), &value)
 		}
 		if !t.canSkip(col, &value) {
-			colIDs = append(colIDs, col.ID)
-			row = append(row, value)
+			buffer.ColIDs = append(buffer.ColIDs, col.ID)
+			buffer.Row = append(buffer.Row, value)
 		}
-		rowToCheck = append(rowToCheck, value)
+		buffer.RowToCheck = append(buffer.RowToCheck, value)
 		if shouldWriteBinlog(sctx.GetSessionVars(), t.meta) && !t.canSkipUpdateBinlog(col, value) {
 			binlogColIDs = append(binlogColIDs, col.ID)
 			binlogOldRow = append(binlogOldRow, oldData[col.Offset])
 			binlogNewRow = append(binlogNewRow, value)
 		}
 	}
 	// check data constraint
-	err = t.CheckRowConstraint(sctx, rowToCheck)
+	err = t.CheckRowConstraint(sctx, buffer.RowToCheck)
```

Comment:
Need to be cautious these slices are not referenced after the `UpdateRecord` call, a possible constraint is to abstarct methods explicitly like:
```
// Only use this when the slice is referenced locally and would NOT be used latter.
func GetRowToCheck() 
// Use this if the slices would be referenced later.
func GetRowToCheckCloned()
```
and comments about the usages.

Using the buffered slice fields directly could make the future maintenance complicated and error-prone.

Reply:
But there is no need for `GetRowToCheckCloned()`. All these slices are temporary variables. Even if a slice is obtained from `GetRowToCheck`, it still can be mistakenly accessed outside its expected lifetime. 
The best option I could think of is to use a clear name that everyone would understand this is a temporary slice.

Reply:
> use a clear name that everyone would understand

Using clearer naming is also feasible. If there are no better hard constraint methods, I believe we need to add comments and warnings in all places where these buffered slices are used, to avoid additionally referencing the input slice and to highlight the related lifetime risks. 

This is actually the most difficult part (or the limitation of Golang). The way mentioned above using functions is essentially an attempt to achieve this goal as much as possible.